rust/connlib/tunnel/src/peer.rs

```diff
@@ -221,11 +220,26 @@ where
         src: &[u8],
         dst: &'b mut [u8],
     ) -> Result<Option<WriteTo<'b>>> {
-        match self.tunnel.lock().decapsulate(None, src, dst) {
+        let mut tunnel = self.tunnel.lock();
+
+        match tunnel.decapsulate(None, src, dst) {
             TunnResult::Done => Ok(None),
             TunnResult::Err(e) => Err(e.into()),
             TunnResult::WriteToNetwork(packet) => {
-                Ok(Some(WriteTo::Network(Bytes::copy_from_slice(packet))))
+                let mut bytes = BytesMut::new();
+                bytes.extend_from_slice(packet);
+
+                // Boringtun requires us to call `decapsulate` repeatedly if it returned `WriteToNetwork`.
+                // However, for the repeated calls, we only need a buffer of at most 148 bytes which we can easily allocate on the stack.
+                let mut buf = [0u8; 148];
+
+                while let TunnResult::WriteToNetwork(packet) =
+                    tunnel.decapsulate(None, &[], &mut buf)
+                {
+                    bytes.extend_from_slice(packet);
```

Comment:
Can we just extend the buffer here instead of sending each packet individually? If we do this, can we assume from the other side a single `read` might give us more than 1 wireguard packet and `decapsulate` might fail.

Reply:
I think to keep this same structure we can [split](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.split) the buffer and call `write` multiple times.

Reply:
> Can we just extend the buffer here instead of sending each packet individually? If we do this, can we assume from the other side a single `read` might give us more than 1 wireguard packet and `decapsulate` might fail.

Given that the API takes a byte slice, I would have assumed that it can (obviously?!) deal with fragments and/or multiple packets. Looking at the boringtun source code though, you seem to be correct that this does in fact expect a single datagram only. What a silly API :roll_eyes: 

> I think to keep this same structure we can [split](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.split) the buffer and call `write` multiple times.

What would you split it by? I think I'd rather extend `WriteTo::Network` to return a list of byte slices.

Reply:
I meant the `split` API will give you a reference to the unused space in the underlying buffer, that way you can pass that and don't have to copy each slice. But nvm! let's optimize copying later, if it's worth it.